### ISSUE TITLE
AM2315 question

### ISSUE THREAD

**Issue by Cabalist** [kizniche/mycodo_python#5]
Hey @kizniche,

This actually calls back to the first comment I made on your codebase.  :)

This code that was in mycodo.py is executed before a reading from AM2315:

```python
            I2C_address = 0x70 + pin // 10
            if GPIO.RPI_REVISION == 2 or GPIO.RPI_REVISION == 3:
                I2C_bus_number = 1
            else:
                I2C_bus_number = 0
            bus = smbus.SMBus(I2C_bus_number)
            bus.write_byte(I2C_address, pin % 10)
            time.sleep(0.1)
        am = AM2315(0x5c, "/dev/i2c-1")
        temperature, humidity, crc_check = am.sense()
```

However compared with the code to test the sensor we have this in ```Test-Sensor-HT-AM2315.py```:
```python
am = AM2315(0x5c, "/dev/i2c-1")

for x in range(10):
    temperature, humidity, crc_check = am.sense()
```

I think that the SMBus is getting setup here.  Could you walk me through what happens there in the extra lines?

**Comment by Cabalist** [comment]
Same code in BMP085-180:
```python         
        I2C_address = 0x70 + pin // 10
        if GPIO.RPI_REVISION == 2 or GPIO.RPI_REVISION == 3:
            I2C_bus_number = 1
        else:
            I2C_bus_number = 0
        bus = smbus.SMBus(I2C_bus_number)
        bus.write_byte(I2C_address, pin % 10)
        time.sleep(0.1)
        press_sensor = BMP085.BMP085()
        temperature = press_sensor.read_temperature()
        pressure = press_sensor.read_pressure()
        altitude = press_sensor.read_altitude()
        # sea_level = sensor.read_sealevel_pressure()
```

**Comment by kizniche** [comment]
SMBus is only used if the I2C multiplexer is in use. If `pin == 0` the multiplexer is not in use:

```python
    elif device == 'AM2315':
        if pin != 0:
            I2C_address = 0x70 + pin // 10
            if GPIO.RPI_REVISION == 2 or GPIO.RPI_REVISION == 3:
                I2C_bus_number = 1
            else:
                I2C_bus_number = 0
            bus = smbus.SMBus(I2C_bus_number)
            bus.write_byte(I2C_address, pin % 10)
            time.sleep(0.1)
        am = AM2315(0x5c, "/dev/i2c-1")
        temperature, humidity, crc_check = am.sense()
```

I should also use the RPi version check to define either /dev/i2c-1 or /dev/i2c-0

I excluded this from the sensor test script because I am assuming the sensor is connected directly to the I2C and not the multiplexer. I have a separate script to test the multiplexer. I suppose I could add a parameter to conduct a reading with and without the multiplexer, but I wanted to keep everything simple for now.

**Comment by kizniche** [comment]
Updated both scripts in commits https://github.com/kizniche/mycodo_python/commit/030e45dd23afe13f0eb53aafdf84bdc6b83446e4 and https://github.com/kizniche/mycodo_python/commit/e5bc602bf78529f56b8ddca3baa39cdc0f687e02

**Comment by Cabalist** [comment]
Perfect!  Thanks!  

What does the write do to the device?  Is it saying "Hey I want a reading"?  I am not seeing it in other documentation for this library

**Comment by kizniche** [comment]
Let's say you have two I2C device with the same address (0x5a). They are connected to the multiplexer (0x70) on channels 1 and 2. You then write to the multiplexer what channel to use (1 - 8). Then you can communicate to 0x5a but it will communicate with the device that you changed the channel to.

**Comment by Cabalist** [comment]
Gotcha.

**Comment by Cabalist** [comment]
I'll be doing a push either tonight or in the next couple days that will have the base level of the hardware access mapped out.  You will definitely need to look over it for confusion on my part like this.  I've broken up all the specific sensor logics in to different functions.  I am right now on the fence on whether it would make sense to have a function called ```read_sensor(sensor_id, session)```  that could take the sensor_id and look at the sensor_registry to determine which and how to read a particular sensor.  That function would return the type of the sensor and it's readings.  

I might commit something that is very rough right now just to illustrate what this looks like.  Would you be OK with that?

**Comment by kizniche** [comment]
Yeah. We're not affecting the main project, so feel free to go wild.

**Comment by Cabalist** [comment]
I am doing my best to write up the new layout for some of this code.  I'm not being good about communicating what I am heading for.

**Comment by kizniche** [comment]
Well, I haven't had many objections so far. As long as you understand what Mycodo currently does, and you don't see a reason why the new code can't eventually support the current feature-set, we're all good.

**Comment by Cabalist** [comment]
Here is what I just wrote up.  Let me know what is missing/wrong and I'll stick it in the Readme

----

Rough summary of Mycodo:

Mycodo at the highest level is a interface and controller for a bunch of sensors stuck in a box.  These are used to control the environment inside that box.  The user needs to setup parameters about the inside of the box and make sure that those parameters are being met.   Visual feedback and alerts are critical as well as consistent logging to verify results.

Rough summary of Mycodo_Python:

I've been calling this mycodo_core in the code.  This handles all sensor access and creates/logs data.  It also generates graphs and reacts to that data but is not the only thing that reads the data.  

-----

Technical Sketch

We start with an array of hardware sensors.  They can be of many types of protocols/types.   We register each one to Mycodo using the WebUI.  By doing this we give each a sensor_id.  This is a unique value and how we refer to that sensor internally.  By doing this we build a Sensor registry.  By selecting the type/model of sensor Mycodo_Python will know how to get data off of it and in what format.

Each sensor is turned on and off by the WebUI.  When turned on the sensors are constantly reading.  Each sensor is running on it’s own thread that is delegated by a controller function internally.  The sensor data is logged to a time series database.  This database acts as the abstraction layer for the rest of the programs access to sensor data.  This avoids contending for a lock on the sensor and inconsistent readings. 

‘Consumers' will be threads or processes that access time series database from the different sensors.  Some examples include:
Log Dumper - Runs every ten minutes and writes the sensor values out to a text file log for easy viewing in Excel or Notepad
Conditional Watchers - Processes that check periodically if particular conditions are met.  If they are it will trigger relays or other actions (email, etc.).  
PID controllers - Similar to above except this monitor feeds into a PID controller to maintain conditions inside the box.
Realtime web graphs - Ajax calls to display current data from inside the box.

Besides programmatically turning on and off sensors and relays, the user needs to turn on and off sensor using the UI.  The web ui will call the mycodo_client to the mycodo_server and instance of an rpyc server.  This way the web server does not need to be running as root and we can control the calls made to the hardware.

The existing SQL database is better known as the 'configuration database’ as it only holds values related to settings for the application.  Mycodo_Python does not currently change any values in it.  There may be some suggestions in the future on optimizing the DB to avoid duplication and ease access to some objects.  

The Mycodo_Python (aka mycodo_core) project runs independently of the web server application.  Currently the two functions of the web server are: 1) consumer of the time series data and 2) configure sensors and relays by writing values to the ‘configuration database’.  These will be maintained as separate pieces from mycodo_python for the moment.

When the web server changes values in the config these values do not need to be ‘reloaded’.   The next time those values are accessed by Mycodo_Python the application will use these new values.

**Comment by kizniche** [comment]
Looks good. I don't see any changes necessary.

**Comment by kizniche** [comment]
Are you familiar with the graphing software I just started using, Highcharts/Highstock, which?

**Comment by Cabalist** [comment]
Yep.  Love em!

**Comment by Cabalist** [comment]
Is there a reason for gnuplot over the highchart version?

**Comment by kizniche** [comment]
I only just discovered Highcharts. It's amazing. The only reason I would want to keep gnuplot is for low-end devices that may be viewing data, as all the processing happens on the server to produce the graph image, instead of having to render it in the browser.

**Comment by Cabalist** [comment]
How lowend are we talking?  Highcharts runs pretty great on an iPhone or android device.  I think the raspberry pi would be on the lowend of the power spectrum these days

**Comment by kizniche** [comment]
I have a 2.5 GHz quad-core phone (S5) and it takes up to 50 seconds to render a Highstock graph with a good amount of data.

**Comment by Cabalist** [comment]
Where is the bottleneck there?  In transferring the data or rendering the data?  Can you save the data sheet locally to check?

**Comment by kizniche** [comment]
One min. while I get some data.

**Comment by kizniche** [comment]
It appears to be mostly the data transfer to the device. 10MB! lol

**Comment by Cabalist** [comment]
I would see if your server could server that compressed.  gzip is great for that type data

**Comment by Cabalist** [comment]
Or lower the resolution of the data ;)

**Comment by kizniche** [comment]
I want all the data points. Aggregation is nice when zoomed out, but I want to potentially see the finest resolution (whatever the measure interval I set).

**Comment by kizniche** [comment]
That's what's nice about gnuplot.. It's a 15KB image :)

**Comment by Cabalist** [comment]
I *believe* that you can serve that with ajax.

**Comment by Cabalist** [comment]
http://www.highcharts.com/stock/demo/lazy-loading

**Comment by kizniche** [comment]
Sweet! Then there's really no reason to keep gnuplot.

**Comment by Cabalist** [comment]
/me immediately drags 2000 lines of code to the trash...

**Comment by kizniche** [comment]
That chart is ridiculous.

**Comment by Cabalist** [comment]
Does it feel weird reducing the code size like this?  You spent a lot of time and sweat making this thing and we are blowing big holes into it.

**Comment by kizniche** [comment]
Now that I found highcharts, I pretty much stopped development using gnuplot. It was inevitable it would be phased out.
